---
title: "Class 13: RNASeq with DESeq2"
author: "Isabel Hui - A16887852"
format: pdf
---

Today we will analyze some RNASeq data from Himes et al. on the effects of dexamethasone (dex), a synthetic glucorticoid steroid on airway smooth muscle cells (ASM).

# Data import 

```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")
```

```{r}
#head(counts)
```

```{r}
#head(metadata)
```

> Q1. How many genes are in this dataset? 

```{r}
nrow(counts)
```

> Q2. How many ‘control’ cell lines do we have? 

```{r}
sum(metadata$dex == "control")
```

```{r}
table(metadata$dex)
```

## Toy Differential Expression Analysis

Calculate the mean per gene count values for all "control" samples (i.e. columns in `counts`), do the same for "treated", and then compare them.

1. Find all "control" values/columns in `counts`.

```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[,control.inds]
```

2. Find the mean per gene across all control columns.

```{r}
control.mean <- apply(control.counts, 1, mean)
```

3. Do the same steps to find the `treated.mean` values.

```{r}
treated.mean <- apply(counts[ ,metadata$dex == "treated"], 1, mean)
```

```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

```{r}
plot(meancounts)
```

It would help to scale since all of the data points seem to be congested.

```{r}
plot(meancounts, log="xy")
```

We most frequently use log2 transformations for this type of data.

```{r}
log2(10/10)
log2(20/10)
log2(10/20)
log2(10/40)
```
These log2 values make the interpretation of "fold-change" a little easier and a rule-of-thumb in the fild is a log2 fold-change of +2 or -2 is where we start to pay attension.

```{r}
log2(40/10)
```

Let's calculate the `log2(fold-change)` and add it to our `meancounts` data.frame.

```{r}
meancounts$log2fc <- log2(meancounts$treated.mean/meancounts$control.mean)
head(meancounts)
```

```{r}
to.rm <- rowSums(meancounts[,1:2]==0) > 0
mycounts <- meancounts[!to.rm,]
```

> Q. How many genes do I have left after this zero count filtering?

```{r}
nrow(mycounts)
```

> Q. How many genes are "up" regulated upon drug treatment at a threshold of +2 log2-fold-change?

1. I need to extract the log2fc values.
2. I need to find those that are above +2.
3. Count them.

```{r}
sum(mycounts$log2fc > 2)
```

> Q. How many genes are "down" regulated upon drug treatment at a threshold of -2 log2-fold-change?

```{r}
sum(mycounts$log2fc < (-2))
```

Wow hold on we are missing the stats here. Is the difference in the mean counts significant?


Let's do this analysis the right way with stats and use the **DESeq** package.

## DESeq Analysis

```{r}
#| message: false
library(DESeq2)
```

The first function that we will use will setup the data in the way (format) DESeq wants it.

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts, 
                       colData = metadata,
                       design = ~dex)
```

The function in the package is called `DESeq()` and we can run it on our `dds` object.

```{r}
dds <- DESeq(dds)
```

I will get the results from dds with the `results()` function:

```{r}
res <- results(dds)
head(res)
```
Make a common overall results figure from this analysis. This is designed to keep out inner biologist and inner stats nerd happy--it is plot fold-change vs. P-value.

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v=c(-2,2), col="gray")
abline(h= -log(0.005), col="gray")
```

Add some color to this plot:

```{r}
mycols <- rep("gray", nrow(res))
mycols[res$log2FoldChange > 2] <- "red"
mycols[res$log2FoldChange < -2] <- "red"
mycols[res$padj > 0.005] <- "gray"


plot(res$log2FoldChange, -log(res$padj), col = mycols)
abline(v=c(-2,2), col="gray")
abline(h= -log(0.005), col="gray")
```

I want to save my results to date out to disc.

```{r}
write.csv(res, file = "myresults.csv")
```

We will pick this up the next class day and add **annotation** (i.e. what are these genes of interest) and do **pathway analysis** (what biology) are they known to be involved with.

```{r}
head(res)
```

## Annotation

I need to translate our gene identifiers "ENSG000..." into gene names that the rest of the world can understand.

To this "annotation" I will use the **AnnotationDbi** package. I can install this with `BiocManager::install()`

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)

columns(org.Hs.eg.db)
```

I will use the `mapIds()` function to "map" my identifiers to those from different databases. I will go between "ENSEMBL" and "SYMBOL" (and then after "GENENAME").

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
       keys = rownames(res),
       keytype = "ENSEMBL",
       column = "SYMBOL")
#head(res)
```

Add "GENENAME":

```{r}
res$genename <- mapIds(org.Hs.eg.db,
       keys = rownames(res),
       keytype = "ENSEMBL",
       column = "GENENAME")
```

Add 

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
       keys = rownames(res),
       keytype = "ENSEMBL",
       column = "ENTREZID")
```
```{r}
head(res)
```

Save our annotated results object.

```{r}
write.csv(res, file = "results_annotated.csv")
```

## Pathway Analysis

Now that we have our results with added annotation we can do some pathway analysis.

Let's use the **gage** package to look for KEGG pathways in our results (genes of interest). I will alos use the **pathview** package to draw little pathway figures.

```{r}
#|message: false
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```

What **gage** wants as input is not my big table/data.frame of results. It just wants a "vector of importance". For RNASeq data like we have this is our log2FC values...

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Now, let's run the gage pathway analysis.

```{r}
#Get the results

keggres = gage(foldchanges, gsets = kegg.sets.hs)
```

What is in this `keggres` object?

```{r}
attributes(keggres)
```

```{r}
head(keggres$less, 3)
```
Let's use the pathview package to look at one of these highlighted KEGG pathways with our genes highlighted.
"hsa05310 Asthma"

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```

![Asthma pathway with my DEGs](hsa05310.pathview.png)
















